src/compiler/checker.ts

```diff
@@ -5983,6 +6038,9 @@ namespace ts {
 
         function getWidenedType(type: Type): Type {
             if (type.flags & TypeFlags.RequiresWidening) {
+                if (type.flags & TypeFlags.PredicateType) {

```

Comment:
I think this is a less common case, so move it below the `undefined`/`null` check.